fabtests/common/shared.c

```diff
@@ -3653,22 +3698,76 @@ int ft_check_atomic(enum ft_atomic_opcodes atomic, enum fi_op op,
 	}
 
 	if (atomic == FT_ATOMIC_COMPARE || atomic == FT_ATOMIC_FETCH) {
-		SWITCH_TYPES(type, CHECK_LOCAL, dst_cpy, res, count, ret);
+		if (opts.iface != FI_HMEM_SYSTEM) {
+			assert(dev_host_res);
+			ret = ft_hmem_copy_from(opts.iface, opts.device,
+						dev_host_res, res,
+						count * datatype_to_size(type));
+			if (ret) {
+				FT_ERR("Failed to copy from atomic buffer\n");
+				return ret;
+			}
+
+			check_res = dev_host_res;
+		} else {
+			check_res = res;
+		}
+		SWITCH_TYPES(type, CHECK_LOCAL, dst_cpy, check_res, count, ret);
 		if (ret) {
 			printf("Data check error on atomic fetch buffer\n");
 			return ret;
 		}
 	}
 
+	if (opts.iface != FI_HMEM_SYSTEM) {
+		assert(dev_host_buf && dev_host_comp);
+		ret = ft_hmem_copy_from(opts.iface, opts.device, dev_host_buf,
+					src, count * datatype_to_size(type));
+		if (ret) {
+			FT_ERR("Failed to copy from atomic buffer\n");
+			return ret;
+		}
+
+		check_buf = dev_host_buf;
+
+		ret = ft_hmem_copy_from(opts.iface, opts.device, dev_host_comp,
+					cmp, count * datatype_to_size(type));
+		if (ret) {
+			FT_ERR("Failed to copy from atomic buffer\n");
+			return ret;
+		}
+
+		check_comp = dev_host_comp;
```

Comment:
No, I think I left it because coverity gets really confused by the unused variables with the atomic cases. But I'm going to refactor this whole check which is looking really messy. I'll push an update when I figure out the other failures.